cmd/setup-etcd-environment/run.go

```diff
@@ -129,53 +164,82 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
 		out = f
 	}
 
-	if runOpts.bootstrapSRV {
-		exportEnv["DISCOVERY_SRV"] = runOpts.discoverySRV
-	} else {
-		exportEnv["NAME"] = etcdName
+	if err := writeEnvironmentFile(exportEnv, out, true); err != nil {
+		return err
 	}
 
-	// enable etcd to run using s390 and s390x. Because these are not officially supported upstream
-	// etcd requires population of environment variable ETCD_UNSUPPORTED_ARCH at runtime.
-	// https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/supported-platform.md
-	arch := runtime.GOARCH
-	if strings.HasPrefix(arch, "s390") {
-		exportEnv["UNSUPPORTED_ARCH"] = arch
+	noExportEnv := map[string]string{
+		"IPV4_ADDRESS":      setupEnv.etcdIP,
+		"WILDCARD_DNS_NAME": fmt.Sprintf("*.%s", setupEnv.opts.discoverySRV),
 	}
-	if err := writeEnvironmentFile(exportEnv, out, true); err != nil {
-		return err
+	if setupEnv.etcdDNS != "" {
+		noExportEnv["DNS_NAME"] = setupEnv.etcdDNS
 	}
+	return writeEnvironmentFile(noExportEnv, out, false)
+}
 
-	return writeEnvironmentFile(map[string]string{
-		"IPV4_ADDRESS":      ip,
-		"DNS_NAME":          dns,
-		"WILDCARD_DNS_NAME": fmt.Sprintf("*.%s", runOpts.discoverySRV),
-	}, out, false)
+func (s *SetupEnv) getExportEnv() (map[string]string, error) {
+	if s.isEtcdScaling() {
+		return s.getOperatorEnv()
+	}
+	// initialize envs used to bootstrap etcd
+	return s.getBootstrapEnv()
 }
 
-func setExportEnv(etcdName, dns string) (map[string]string, error) {
-	exportEnv := make(map[string]string)
-	duration := 10 * time.Second
-	wait.PollInfinite(duration, func() (bool, error) {
-		if _, err := os.Stat("/var/run/secrets/kubernetes.io/serviceaccount/token"); os.IsNotExist(err) {
-			glog.Errorf("serviceaccount failed: %v", err)
-			return false, nil
-		}
-		return true, nil
-	})
+// getOperatorEnv returns a maping of ENV variables required to scale etcd via operator
+func (s *SetupEnv) getOperatorEnv() (map[string]string, error) {
+	env := make(map[string]string)
+	env["NAME"] = s.etcdName
+
+	memberList := s.getEtcdMemberList()
+	env["INITIAL_CLUSTER"] = strings.Join(memberList, ",")
```

Comment:
this ends up being not an array but a string: "blah, blah, blah" ?

wondering why this is a string and not an array (prob a good reason, but would like to understand)

Reply:
This is typical of the INITIAL_CLUSTER environment variable to have a comma separated string of member list. For instance, we build the INITIAL_CLUSTER when we restore a backup we set the INITIAL_CLUSTER as:
echo $INITIAL_CLUSTER
etcd-member-ip-10-0-131-52.ec2.internal=https://etcd-1.alpatel-test-3.devcluster.openshift.com:2380,etcd-member-ip-10-0-157-131.ec2.internal=https://etcd-1.alpatel-test-3.devcluster.openshift.com:2380,etcd-member-ip-10-0-164-217.ec2.internal=https://etcd-2.alpatel-test-3.devcluster.openshift.com:2380

See https://docs.openshift.com/container-platform/4.1/backup_and_restore/disaster_recovery/scenario-2-restoring-cluster-state.html (section 1.c)

Reply:
Sure so ETCD_INITIAL_CLUSTER [1] is a comma-delimited string [2] that accepts peers in the following format `<name>=<peerURL>`.

[1] https://github.com/etcd-io/etcd/blob/a621d807f061e1dd635033a8d6bc261461429e27/Documentation/op-guide/configuration.md#--initial-cluster

[2] https://github.com/etcd-io/etcd/blob/a65e4d3357c260504d47f08a433b355e36dbcb51/embed/config.go#L198

Reply:
ahhhh TIL.. thank you both for explaining and linking! :)